### PR TITLE
Versioning API (WIP)

### PR DESCRIPTION
**Pull request by pfrazee**
This PR is a work-in-progress. It will add version checkpoints, which Hyperdrive's API can create and list, and use to find specific versions of files in its history.

I've written a new message-type in the schema, and updated the readme with the API that I propose for this. I need the changes to `list()` to be merged (https://github.com/mafintosh/hyperdrive/pull/99) before I can finish updating the API (`list` will take a 'version' option).

Let me know if you think anything should change!

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -117,6 +117,15 @@ Otherwise it is a 32 byte hash.
 
 Boolean whether archive is live. `true` by default. Note that its only populated after archive.open(cb) has been fired.
 
+#### `archive.version`
+
+The label of the most recent checkpoint.
```
Comment (juliangruber) [inline README.md]: maybe document that this is only available after it has finished opening?

File: README.md [hunk 2]
```diff
@@ -117,6 +117,15 @@ Otherwise it is a 32 byte hash.
...
+
+#### `archive.semver`
+
+The label of the most recent semantic-version checkpoint.
```
Comment (juliangruber) [inline README.md]: the semver thing sounds like an addition that could be useful too but might better be discussed in a separate pull request?

File: README.md [hunk 3]
```diff
@@ -138,6 +147,25 @@ archive.append('hello.txt', function () {
...
+Write a version checkpoint to the archive, marking the state for future reference. Version includes:
+
+```js
+{
+  label: '1.0.0', // can be any string, but semantic versions are recommended
+  message: 'The first version!' // an optional string describing the checkpoint
+}
+```
+
+If the label has been used before, Hyperdrive will error.
```
Comment (juliangruber) [inline README.md]: to me `label` isn't a term for an unique datum, maybe let's use `version` instead?

**Comment by karissa** [comment]
What's the difference between version and semver?

**Comment by karissa** [comment]
That was my first question. Maybe having a different name for version would be helpful or removing the "semver" property would be less confusion. Wasn't entirely clear to me at first

File: README.md (hunk 1 above)
Comment (pfrazee) [inline README.md]: Yep

File: README.md (hunk 3 above)
Comment (pfrazee) [inline README.md]: I think `version` might be fine in this API call, but I think the message schema should still use `label`. This is the schema:

```
message Checkpoint {
  required string label = 1;
  required uint64 timestamp = 2;
  optional string message = 3;
}
```

`version` wouldn't make sense. It should be `id` or `name`.

**Comment by pfrazee** [comment]
I'll remove `.semver` for now

**Comment by pfrazee** [comment]
Pushed some updates based on the comments

**Comment by juliangruber** [comment]
hm yeah i get your point about using `label`, but then i think we shouldn't call it `version` at all. label is a broader term that works for more use cases.

**Comment by pfrazee** [comment]
That is a possibility: just remove all 'version' semantics entirely from hyperdrive. They're purely checkpoints that add labels to points in the history, and they can be used downstream to implement versions.

In that case, would we have `.checkpoint()` enforce semver monotonicity? We could, I suppose, and still remove most other semantics about versions.

**Comment by joehand** [comment]
> That is a possibility: just remove all 'version' semantics entirely from hyperdrive. They're purely checkpoints that add labels to points in the history, and they can be used downstream to implement versions.

👍  on this. After hearing about versions from RDA, there are 1001 way scientists/publishers/etc. want to do versions. If we allow a checkpoint function and some way to get all checkpoints (so they can enforce whatever downstream) that would be ideal.

**Comment by joehand** [comment]
> That is a possibility: just remove all 'version' semantics entirely from hyperdrive. They're purely checkpoints that add labels to points in the history, and they can be used downstream to implement versions.

👍  on this. After hearing about versions from RDA, there are 1001 way scientists/publishers/etc. want to do versions. If we allow a checkpoint function and some way to get all checkpoints (so they can enforce whatever downstream) that would be ideal.

**Comment by pfrazee** [comment]
Ok that's cool with me. Do we want to consider calling it a "tag" instead
of a "checkpoint," to match the semantics from Git ?

**Comment by joehand** [comment]
Right now we have these ideas:

* Live (latest)
* Snapshot
* Checkpoint (snapshot + label/message)

That seems to make sense to me.

**Comment by pfrazee** [comment]
What's a snapshot?

**Comment by juliangruber** [comment]
@pfrazee like a checkpoint but without a label. just an archive at one point in time



> That is a possibility: just remove all 'version' semantics entirely from hyperdrive. They're purely checkpoints that add labels to points in the history, and they can be used downstream to implement versions.
> 
> In that case, would we have .checkpoint() enforce semver monotonicity? We could, I suppose, and still remove most other semantics about versions.

I don't fully understand this, are you saying to remove versions or to add semver?

**Comment by karissa** [comment]
Let's try to keep it as simple as possible. Enforcing semver monotonicity sounds like a feature that could be nice but some people might want to turn it off. So maybe handle in a separate PR?

Everything looks great, I'd say the names matter less. 

Checkpoint is more like a commit (+metadata).
Snapshot is like a version of the repo, doesn't have metadata. It's just a hash. I'd say its more for internal use or "oopsie" moments where you want to revert. Kinda like google docs or dropbox versions. Let me know if this isn't right @joehand

**Comment by joehand** [comment]
> Checkpoint is more like a commit (+metadata).
Snapshot is like a version of the repo, doesn't have metadata. It's just a hash. I'd say its more for internal use or "oopsie" moments where you want to revert. Kinda like google docs or dropbox versions. Let me know if this isn't right @joehand

Exactly!

**Comment by pfrazee** [comment]
The version semantics are removed

**Comment by karissa** [comment]
nice! ltm

File: schema.proto [hunk 4]
```diff
@@ -34,3 +34,9 @@ message Entry {
 //}
 
 // missing: truncate, chown, chmod,
+
+message Checkpoint {
+  required string name = 1;
+  required uint64 timestamp = 2;
+  optional string message = 3;
```
Comment (juliangruber) [inline schema.proto]: what about `name`, `timestamp` and `description`?

Comment (pfrazee) [inline schema.proto]: Sure

**Comment by karissa** [comment]
this + history api will make me so happy

**Comment by karissa** [comment]
itching to merge

**Comment by pfrazee** [comment]
@karissa :) I was waiting for #99 to merge before writing this PR's code, but I'll go ahead and get that done now

**Comment by pfrazee** [comment]
Checkpoint() and checkpoints() are now implemented. DONT MERGE BEFORE #99 IS MERGED! There are new messages, with new numbers, assigned in both PRs. (Unlink in the other, Checkpoint in this.) Currently they both have the same message number, and that no good.

**Comment by pfrazee** [comment]
Todo: update checkpoint msg to hyperdrive-encoding

**Comment by pfrazee** [comment]
Todo: add a hash of the content using https://github.com/mafintosh/hypercore/blob/master/lib/feed.js#L152

**Comment by pfrazee** [comment]
I rebased this PR to latest on master. Here's the message schema I used, which will need to be moved into hyperdrive-encodings:

```
message Checkpoint {
  required string name = 1;
  required uint64 timestamp = 2;
  optional string description = 3;
} 
```

**Comment by mafintosh** [comment]
See .checkout in 8
